snuba/datasets/groups.py

```diff
@@ -0,0 +1,119 @@
+
+from snuba.datasets import TimeSeriesDataset
+from snuba.datasets.dataset_schemas import DatasetSchemas
+from snuba.datasets.factory import get_dataset
+from snuba.datasets.schemas.join import (
+    JoinConditionExpression,
+    JoinCondition,
+    JoinedSchema,
+    JoinStructure,
+    JoinType,
+    SchemaJoinedSource,
+)
+from snuba.query.extensions import (
+    PerformanceExtension,
+    ProjectExtension,
+    QueryExtension,
+)
+from snuba.query.query import AliasedColumn
+from snuba.query.timeseries import TimeSeriesExtension
+
+
+class Groups(TimeSeriesDataset):
+    """
+    Experimental dataset that provides Groups data joined with
+    the events table.
+    """
+
+    EVENTS_ALIAS = "events"
+    GROUPS_ALIAS = "groups"
+
+    QUALIFIED_COLUMN_REGEX = re.compile(r"^([a-zA-Z_][a-zA-Z0-9_]*)\.([a-zA-Z0-9_\.]+)$")
+
+    def __init__(self) -> None:
+        self.__grouped_message = get_dataset("groupedmessage")
+        self.__events = get_dataset("events")
+
+        join_structure = JoinStructure(
+            left_source=SchemaJoinedSource(
+                self.GROUPS_ALIAS,
+                self.__grouped_message.get_dataset_schemas().get_read_schema(),
+            ),
+            right_source=SchemaJoinedSource(
+                self.EVENTS_ALIAS,
+                self.__events.get_dataset_schemas().get_read_schema(),
+            ),
+            mapping=[
+                JoinCondition(
+                    left=JoinConditionExpression(table_alias="groups", column="project_id"),
+                    right=JoinConditionExpression(table_alias="events", column="project_id"),
+                ),
+                JoinCondition(
+                    left=JoinConditionExpression(table_alias="groups", column="id"),
+                    right=JoinConditionExpression(table_alias="events", column="group_id"),
+                ),
+            ],
+            join_type=JoinType.LEFT,
+        )
+
+        schema = JoinedSchema(join_structure)
+        dataset_schemas = DatasetSchemas(
+            read_schema=schema,
+            write_schema=None,
+        )
+        super().__init__(
+            dataset_schemas=dataset_schemas,
+            processor=None,
+            default_topic=None,
+            time_group_columns={
+                'events.time': 'events.timestamp',
+            },
+        )
+
+    def default_conditions(self):
+        return [
+            ('events.deleted', '=', 0),
+            ('groups.record_deleted', '=', 0),
+        ]
+
+    def __parse_qualified_column(self, column_name: str) -> AliasedColumn:
+        # TODO: This logic should be general and in the Query class.
+        # cannot do that yet since the column processing methods like
+        # column_expr do not have access to the Query yet.
+        match = self.QUALIFIED_COLUMN_REGEX.match(column_name)
+        if not match or not match[1] in [self.EVENTS_ALIAS, self.GROUPS_ALIAS]:
+            return (None, column_name)
```

Comment:
Not sure I understand the `not match[1] in [self.EVENTS_ALIAS, self.GROUPS_ALIAS]` part of this expression. Wouldn't this be an error condition if somebody provided an invalid table alias?

Reply:
The goal is to  still allow expressions like tags[], tags_key and so on to be used without alias to avoid making the events dataset know about the concept of alias. 
But, thinking over it 2 days later, I don't think it is a bad idea for the events dataset to know that a column name may have an alias in front. Will change it that way.